File: examples/agent-sdk/js/first-turn.mjs
// Run one Jan Agent task from Node.js and stream the answer.
//
// Starts `jan cli agent run` with `--output-format stream-json`, prints the
// answer as it streams, and reads the terminal `result` record. No dependencies.
//
//     node first-turn.mjs "Explain what a mutex is in one sentence."
//
// Set JAN_BIN to use a jan binary that is not on PATH, and JAN_MODEL to pick a model.

import { spawn } from 'node:child_process'
import { mkdtempSync, rmSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { createInterface } from 'node:readline'

const jan = process.env.JAN_BIN ?? 'jan'
const task = process.argv[2] ?? 'Reply with one short sentence: what is Jan Agent?'

// A fresh, empty project: the agent's own file and shell tools are confined to
// it, so an example cannot touch the directory you ran it from.
const project = mkdtempSync(join(tmpdir(), 'jan-first-turn-'))
const args = ['cli', 'agent', 'run', '--project', project, '--output-format', 'stream-json']
if (process.env.JAN_MODEL) args.push('--model', process.env.JAN_MODEL)
args.push(task)

const run = spawn(jan, args, { stdio: ['ignore', 'pipe', 'inherit'] })
const exited = new Promise((resolve) => run.on('close', resolve))

let result = null
for await (const line of createInterface({ input: run.stdout })) {
  const record = JSON.parse(line)
  if (record.type === 'init') {
    console.error(`[session ${record.session_id} on ${record.model}]`)
  } else if (record.type === 'token') {
    process.stdout.write(record.text)
  } else if (record.type === 'result') {
    result = record
  }
}
const code = await exited
rmSync(project, { recursive: true, force: true })

console.log()
if (result === null) {
  console.error(`jan exited with ${code} before printing a result`)
  process.exit(1)
}
if (result.is_error) {
  console.error(`run failed: ${result.error.code}: ${result.error.message}`)
  process.exit(1)
}
console.error(`[${result.stop_reason} after ${result.num_turns} turn(s)]`)

File: examples/adk/python/first_turn.py
"""Run one Jan Agent task from Python and stream the answer.

Starts `jan cli agent run` with `--output-format stream-json`, prints the answer
as it streams, and reads the terminal `result` record. Standard library only.

    python3 first_turn.py "Explain what a mutex is in one sentence."

Set JAN_BIN to use a jan binary that is not on PATH, and JAN_MODEL to pick a model.
"""

import json
import os
import subprocess
import sys
import tempfile

JAN = os.environ.get("JAN_BIN", "jan")


def main() -> int:
    task = sys.argv[1] if len(sys.argv) > 1 else "Reply with one short sentence: what is Jan Agent?"
    # A fresh, empty project: the agent's own file and shell tools are confined
    # to it, so an example cannot touch the directory you ran it from.
    with tempfile.TemporaryDirectory() as project:
        cmd = [JAN, "cli", "agent", "run", "--project", project, "--output-format", "stream-json"]
        if os.environ.get("JAN_MODEL"):
            cmd += ["--model", os.environ["JAN_MODEL"]]
        cmd.append(task)

        run = subprocess.Popen(cmd, stdout=subprocess.PIPE, text=True)
        result = None
        for line in run.stdout:
            record = json.loads(line)
            if record["type"] == "init":
                print(f"[session {record['session_id']} on {record['model']}]", file=sys.stderr)
            elif record["type"] == "token":
                print(record["text"], end="", flush=True)
            elif record["type"] == "result":
                result = record
        run.wait()

    print()
    if result is None:
        print(f"jan exited with {run.returncode} before printing a result", file=sys.stderr)
        return 1
    if result["is_error"]:
        print(f"run failed: {result['error']['code']}: {result['error']['message']}", file=sys.stderr)
        return 1
    print(f"[{result['stop_reason']} after {result['num_turns']} turn(s)]", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
